services/src/main/java/org/keycloak/truststore/TruststoreBuilder.java

```diff
@@ -0,0 +1,246 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.truststore;
+
+import org.jboss.logging.Logger;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Builds a system-wide truststore from the given config options.
+ */
+public class TruststoreBuilder {
+
+    public static final String SYSTEM_TRUSTSTORE_KEY = "javax.net.ssl.trustStore";
+    public static final String SYSTEM_TRUSTSTORE_PASSWORD_KEY = "javax.net.ssl.trustStorePassword";
+    public static final String SYSTEM_TRUSTSTORE_TYPE_KEY = "javax.net.ssl.trustStoreType";
+    private static final String CERT_PROTECTION_ALGORITHM_KEY = "keystore.pkcs12.certProtectionAlgorithm";
+    public static final String DUMMY_PASSWORD = "keycloakchangeit"; // fips length compliant dummy password
+    static final String PKCS12 = "PKCS12";
+
+    private static final Logger LOGGER = Logger.getLogger(TruststoreBuilder.class);
+
+    public static void setSystemTruststore(String[] truststores, boolean trustStoreIncludeDefault, String dataDir) {
+        KeyStore truststore = createMergedTruststore(truststores, trustStoreIncludeDefault);
+
+        // save with a dummy password just in case some logic that uses the system properties needs to have one
+        File file = saveTruststore(truststore, dataDir, DUMMY_PASSWORD.toCharArray());
+
+        // finally update the system properties
+        System.setProperty(TruststoreBuilder.SYSTEM_TRUSTSTORE_KEY, file.getAbsolutePath());
+        System.setProperty(TruststoreBuilder.SYSTEM_TRUSTSTORE_TYPE_KEY, PKCS12);
+        System.setProperty(TruststoreBuilder.SYSTEM_TRUSTSTORE_PASSWORD_KEY, DUMMY_PASSWORD);
+    }
+
+    static File saveTruststore(KeyStore truststore, String dataDir, char[] password) {
+        File file = new File(dataDir, "keycloak-truststore.p12");
+        file.getParentFile().mkdirs();
+        try (FileOutputStream fos = new FileOutputStream(file)) {
+            // this should inhibit the use of encryption in storing the certs
+            // it's of course not concurrency safe, but it should only be run at startup
+            String oldValue = System.setProperty(CERT_PROTECTION_ALGORITHM_KEY, "NONE");
+            truststore.store(fos, password);
```

Comment:
That is the assumption yes - all of the certs obtained should not have been encrypted.  The only exception to this is if a user overrides the javax.net.ssl properties to specify a truststore that has encrypted certs.  However since they are supposed to be public in their usage, there still shouldn't be a concern.  

Are you asking about this in reference to a potential non-x509 certificate, which may not be public?

Reply:
> That is the assumption yes - all of the certs obtained should not have been encrypted. The only exception to this is if a user overrides the javax.net.ssl properties to specify a truststore that has encrypted certs. However since they are supposed to be public in their usage, there still shouldn't be a concern.

Makes sense. 

> Are you asking about this in reference to a potential non-x509 certificate, which may not be public?

Not really. The explanation above is sufficient for me.